neuralhydrology/datasetzoo/basedataset.py

```diff
@@ -291,6 +298,9 @@ def _load_or_create_xarray_dataset(self) -> xarray.Dataset:
                 # add columns from dataframes passed as additional data files
                 df = pd.concat([df, *[d[basin] for d in self.additional_features]], axis=1)
 
+                # if target variables are missing for basin, add empty column to still allow predictions to be made
+                df = self._add_missing_targets(df)
```

Comment:
Even though it would not cause harm to allow this during training, I would suggest to add an

```python
if self.is_train:
    df = self._add_missing_targets(df)
```

What would happen during training currently is that you could specify a basin, which has no obs values at all. For these basins you would add a NaN column and the `validate_samples` function would then remove all samples of this basin from the possible training samples (hence "no harm"). But it seems a bit like an odd behavior.

Reply:
Yes, I agree that makes sense to not have this be allowed for training. I've added in `if not self.is_train:` to check that

Reply:
> Now what you say is that it would be nice to call a function (from e.g. the shell) that evaluates a model on all basins in a separate basin file, but instead of evaluating them and trying to compute metrics, it only gets the predictions and stores them to disk? But isn't that what the Tester class does?

Yes, the Tester class does basically do just what I was describing and already accommodates no metrics being calculated. Apologies if it was a little unclear what I was describing. I was suggesting it would be convenient for our use case (and was wondering if would be for others generally) is something that is built-in more directly for model predictions on a basin file that is separate from the train, validate, and test dataset. So, we want to output model predictions without overwriting the `validation_results.p` or `test_results.p` files. One could certainly do that in a notebook with calling the forward pass of the network like you suggest (maybe that would even be the best option). But I think I often see a high level `predict` mode function in the different python ML packages that operates outside the `train, validation, and test` datasets. That's what I was trying to get it in my suggestion. So, one thing that came to mind would be to create a 4th allowable dataset partition (call it "predict" of similar) that one could call the `Tester` class on. The 2nd option that came to mind was to create a `Predictor` class which I think would end up being basically a slimmed down `Tester` class (since no handling of observations, train/validation/test modes, or metrics would be needed) that doesn't require the `period` to be `train`, `validation`, or `test`.  

I think we'll end up coding up something to make predictions outside of the train/validation/test periods, so if helpful, I could circle back at that point to share our code to see if it would be of interest to develop into a PR that gets merged.